compiler/passes/scopeResolve.cpp

```diff
@@ -725,43 +758,80 @@ void resolveUnresolvedSymExprs(BaseAST* inAst) {
    }
 }
 
-static void warnUnstableClassType(SymExpr* se) {
-  if (TypeSymbol* ts = toTypeSymbol(se->symbol())) {
-    if (isClass(ts->type)) {
-      if (se->getModule()->modTag == MOD_USER) {
+static bool isStableClassType(Type* t) {
+  bool ok = false;
+
+  TypeSymbol* ts = t->symbol;
+
+  if (isClass(t)) {
+    // Always consider locale type unmanaged
+    if (ts->type == dtLocale)
+      ok = true;
+    // Always consider ddata type unmanaged
+    if (ts->hasFlag(FLAG_DATA_CLASS))
+      ok = true;
+    // Something with "no object" flag isn't really an object anyway
+    if (ts->hasFlag(FLAG_NO_OBJECT))
+      ok = true;
+  }
+
+  return ok;
+}
+
+static bool callSpecifiesClassKind(CallExpr* call) {
+  return (call->isPrimitive(PRIM_TO_UNMANAGED_CLASS) ||
+          call->isPrimitive(PRIM_TO_BORROWED_CLASS) ||
+          call->isNamed("_to_borrowed") ||
+          call->isNamed("_to_unmanaged") ||
+          call->isNamed("_owned") ||
+          call->isNamed("_shared") ||
+          call->isNamed("Owned") ||
+          call->isNamed("Shared"));
+}
+
+// Returns the expr resulting, either se or a call containing it
```

Comment:
What if this function returns the resulting Type* directly?

Either way, I'd change the above comment to clarify the intention. For example:

"Handle the case where se->symbol() is "unstable" i.e. an undecorated class type. With --warn-unstable, issue a warning. With --default-unmanaged, wrap it in a _toUnmanaged call. Return either se or a call _toUnmanaged(se)".